src/async_impl/client.rs

```diff
@@ -395,29 +396,42 @@ impl Client {
         ) = request::pieces(req);
 
         let mut headers = self.inner.headers.clone(); // default headers
-        headers.extend(user_headers.iter());
+        for (key, value) in user_headers.iter() {
+            headers.insert(key, value.clone());
+        }
 
         if self.inner.gzip &&
-            !headers.has::<AcceptEncoding>() &&
-            !headers.has::<Range>() {
-            headers.set(AcceptEncoding(vec![qitem(Encoding::Gzip)]));
+            !headers.contains_key(ACCEPT_ENCODING) &&
+            !headers.contains_key(RANGE) {
+            headers.insert(ACCEPT_ENCODING, HeaderValue::from_static("gzip"));
         }
 
         let uri = to_uri(&url);
-        let mut req = ::hyper::Request::new(method.clone(), uri.clone());
-        *req.headers_mut() = headers.clone();
-        let body = body.map(|body| {
+        let (body, req) = match body.map(|body| {
```

Comment:
I believe this is the cause of the problems. It's mapping on `body`, and only if there was a body, does it create a `Request` to be stored as `Some(req)`.

You could change this to just map the `body` into the type to be passed to `hyper` (which if it's empty, there's `hyper::Body::empty()`), and then always construct a `Request` with the mapped body.

Reply:
Yes!! This worked :smile: